File: fix_json.js
// Script Node.js sửa lỗi file JSON ngân hàng câu hỏi trắc nghiệm LSD
// Chạy: node fix_json.js

const fs = require('fs');
const path = require('path');

const INPUT = path.join(__dirname, 'tracnghiem_LSD_chuong_gemini.json');
const OUTPUT = path.join(__dirname, 'tracnghiem_LSD_chuong_gemini_fixed.json');
const REPORT = path.join(__dirname, 'bao_cao_loi.md');

console.log('Đang đọc file...');
const raw = fs.readFileSync(INPUT, 'utf-8');
const data = JSON.parse(raw);
let questions = data.questions;
console.log(`Tổng câu hỏi: ${questions.length}`);

const report = [];
report.push('# BÁO CÁO PHÂN TÍCH & SỬA LỖI\n\n');
report.push(`**Tổng câu gốc:** ${questions.length}\n\n`);

// ============================================================
// LỖI 5: OCR - chQ -> chủ, cQa -> của, phQ -> phủ
// ============================================================
const ocrMap = [['chQ','chủ'],['cQa','của'],['phQ','phủ'],['PhQ','Phủ']];
let ocrFixes = 0;
questions.forEach((q, i) => {
  ocrMap.forEach(([old, nw]) => {
    if (q.question.includes(old)) { q.question = q.question.replaceAll(old, nw); ocrFixes++; }
    if (q.answer.includes(old)) { q.answer = q.answer.replaceAll(old, nw); ocrFixes++; }
    q.options = q.options.map(o => {
      if (o.includes(old)) { ocrFixes++; return o.replaceAll(old, nw); }
      return o;
    });
  });
});
console.log(`Sửa OCR: ${ocrFixes}`);
report.push(`## LỖI 5: OCR\n- Đã sửa: **${ocrFixes}** chỗ\n\n`);

// ============================================================
// LỖI 1b: Xóa số dính ở đầu câu
// ============================================================
let digitFixes = 0;
const digitFixDetails = [];
questions.forEach((q, i) => {
  const text = q.question.trim();
  if (/^\d+\s+[A-ZĐ]/.test(text)) {
    const oldText = text.substring(0, 80);
    q.question = text.replace(/^\d+\s+/, '');
    digitFixes++;
    digitFixDetails.push(`  - [${i}] "${oldText}" → "${q.question.substring(0, 80)}"`);
  }
});
console.log(`Xóa số dính: ${digitFixes}`);
report.push(`## LỖI 1b: Số dính ở đầu\n- Đã sửa: **${digitFixes}** câu\n`);
digitFixDetails.forEach(d => report.push(d + '\n'));
report.push('\n');

// ============================================================
// LỖI 4: Sửa thứ tự đáp án A,B,C,D
// ============================================================
const expectedLetters = ['A','B','C','D'];
let orderFixes = 0;
const orderFixDetails = [];
questions.forEach((q, i) => {
  const opts = q.options;
  const letters = opts.map(o => { const m = o.trim().match(/^([A-Z])\./); return m ? m[1] : '?'; });
  const exp = expectedLetters.slice(0, opts.length);
  if (JSON.stringify(letters) !== JSON.stringify(exp)) {
    orderFixDetails.push(`  - [${i}] ${letters.join(',')} → ${exp.join(',')}: ${q.question.substring(0, 60)}`);
    opts.forEach((opt, j) => {
      const cl = expectedLetters[j];
      const newOpt = opt.trim().replace(/^[A-Z]\.\s/, cl + '. ');
      if (opt.trim() === q.answer.trim()) q.answer = newOpt;
      q.options[j] = newOpt;
    });
    orderFixes++;
  }
});
console.log(`Sửa thứ tự: ${orderFixes}`);
report.push(`## LỖI 4: Sai thứ tự phương án\n- Đã sửa: **${orderFixes}** câu\n`);
orderFixDetails.forEach(d => report.push(d + '\n'));
report.push('\n');

// ============================================================
// LỖI 6: Loại bỏ câu trùng lặp
// ============================================================
// Đếm trước
const countMap = {};
questions.forEach(q => {
  const t = q.question.trim();
  countMap[t] = (countMap[t] || 0) + 1;
});
const dupeGroups = Object.entries(countMap).filter(([_, c]) => c > 1).sort((a, b) => b[1] - a[1]);

const seen = new Set();
const uniqueQ = [];
let removed = 0;
questions.forEach(q => {
  const t = q.question.trim();
  if (!seen.has(t)) { seen.add(t); uniqueQ.push(q); } else { removed++; }
});
console.log(`Loại trùng: ${removed} (${dupeGroups.length} nhóm)`);

report.push(`## LỖI 6: Câu hỏi lặp lại\n`);
report.push(`- Nhóm lặp: **${dupeGroups.length}**\n`);
report.push(`- Bản sao đã xóa: **${removed}**\n`);
report.push(`- Top câu lặp nhiều nhất:\n`);
dupeGroups.slice(0, 30).forEach(([text, count]) => {
  report.push(`  - [${count} lần] ${text.substring(0, 100)}\n`);
});
report.push('\n');

// Cập nhật data
data.questions = uniqueQ;
data.totalQ = uniqueQ.length;
const sc = {};
uniqueQ.forEach(q => { sc[q.section] = (sc[q.section] || 0) + 1; });
data.sections.forEach(s => { s.count = sc[s.key] || 0; });

// ============================================================
// THỐNG KÊ LỖI CÒN LẠI (cần sửa thủ công)
// ============================================================

// Lỗi 1: Câu bắt đầu chữ thường
const lowerStart = [];
uniqueQ.forEach((q, i) => {
  const text = q.question.trim();
  if (text && /^[a-zàáảãạăắằẳẵặâấầẩẫậèéẻẽẹêếềểễệìíỉĩịòóỏõọôốồổỗộơớờởỡợùúủũụưứừửữựỳýỷỹỵđ]/.test(text)) {
    lowerStart.push({idx: i, text: text.substring(0, 150)});
  }
});
report.push(`## LỖI 1: Mất chữ đầu (CẦN SỬA THỦ CÔNG)\n`);
report.push(`- Tổng: **${lowerStart.length}** câu\n`);
report.push(`> ⚠️ Không thể tự đoán chữ/đoạn bị thiếu\n\n`);
lowerStart.forEach(({idx, text}) => {
  report.push(`- \`[Câu ${idx}]\` ${text}\n`);
});
report.push('\n');

// Lỗi 3: Thiếu đáp án
const missingOpts = [];
uniqueQ.forEach((q, i) => {
  if (q.options.length < 4) {
    missingOpts.push({idx: i, count: q.options.length, text: q.question.substring(0, 100)});
  }
});
report.push(`## LỖI 3: Thiếu phương án (CẦN SỬA THỦ CÔNG)\n`);
report.push(`- Tổng: **${missingOpts.length}** câu\n`);
report.push(`> ⚠️ Cần bổ sung đáp án từ nguồn gốc\n\n`);
missingOpts.forEach(({idx, count, text}) => {
  report.push(`- \`[Câu ${idx}]\` Chỉ có **${count}** đáp án: ${text}\n`);
});
report.push('\n');

// Lỗi 2: Đáp án bị cắt
const cutWords = [' và',' của',' tư',' cho',' để',' là',' về',' theo',' trong',' với',' đã',' một',' các',' có',' được',' do',' từ',' lên',' ra',' vào',' mà'];
const cutAnswers = [];
uniqueQ.forEach((q, i) => {
  q.options.forEach((opt, j) => {
    const clean = opt.trim();
    const last = clean[clean.length - 1];
    if (last && !'.)?!…;:"\''.includes(last)) {
      for (const cw of cutWords) {
        if (clean.endsWith(cw)) {
          cutAnswers.push({idx: i, letter: String.fromCharCode(65+j), ending: clean.slice(-70), qText: q.question.substring(0, 80)});
          break;
        }
      }
    }
  });
});
report.push(`## LỖI 2: Đáp án bị cắt xén (CẦN SỬA THỦ CÔNG)\n`);
report.push(`- Tổng: **${cutAnswers.length}** đáp án\n`);
report.push(`> ⚠️ Cần bổ sung phần bị cắt từ nguồn gốc\n\n`);
cutAnswers.forEach(({idx, letter, ending, qText}) => {
  report.push(`- \`[Câu ${idx}]\` Đáp án **${letter}**: ...${ending}\n`);
  report.push(`  - Câu hỏi: ${qText}\n`);
});
report.push('\n');

// ============================================================
// TỔNG KẾT
// ============================================================
report.push('---\n\n## TỔNG KẾT\n\n');
report.push('| Hạng mục | Số lượng |\n');
report.push('|----------|----------|\n');
report.push(`| Câu gốc | **${questions.length}** |\n`);
report.push(`| Sau khi sửa & loại trùng | **${uniqueQ.length}** |\n`);
report.push(`| Đã loại câu trùng | **${removed}** |\n`);
report.push(`| Sửa OCR | **${ocrFixes}** |\n`);
report.push(`| Xóa số dính đầu | **${digitFixes}** |\n`);
report.push(`| Sửa thứ tự đáp án | **${orderFixes}** |\n\n`);

report.push('### Phân bố theo chương:\n');
data.sections.forEach(s => {
  report.push(`- **${s.label}**: ${s.count} câu\n`);
});
report.push('\n### Lỗi cần sửa thủ công:\n');
report.push(`- Mất chữ đầu: **${lowerStart.length}** câu\n`);
report.push(`- Thiếu đáp án: **${missingOpts.length}** câu\n`);
report.push(`- Đáp án bị cắt: **${cutAnswers.length}** đáp án\n`);

// ============================================================
// GHI FILE
// ============================================================
fs.writeFileSync(OUTPUT, JSON.stringify(data, null, 2), 'utf-8');
console.log(`\nĐã ghi: ${OUTPUT}`);

fs.writeFileSync(REPORT, report.join(''), 'utf-8');
console.log(`Đã ghi: ${REPORT}`);

console.log(`\n=== KẾT QUẢ ===`);
console.log(`${questions.length} → ${uniqueQ.length} câu (loại ${removed} trùng)`);
console.log(`OCR: ${ocrFixes} | Số dính: ${digitFixes} | Thứ tự: ${orderFixes}`);
console.log(`Cần sửa thủ công: chữ thường=${lowerStart.length}, thiếu ĐA=${missingOpts.length}, cắt xén=${cutAnswers.length}`);

File: fix_json.py
# -*- coding: utf-8 -*-
"""
Script phân tích và sửa lỗi file JSON ngân hàng câu hỏi trắc nghiệm Lịch sử Đảng.
Xử lý 6 loại lỗi:
1. Mất chữ đầu / Khuyết đoạn đầu
2. Đáp án bị cắt xén
3. Thiếu số lượng phương án
4. Sai format thứ tự phương án
5. Lỗi OCR (chQ, cQa, phQ)
6. Câu hỏi lặp lại
"""
import json
import re
import os
from collections import Counter

# === CẤU HÌNH ===
INPUT_FILE = r'd:\Download\Thi\LSD\tracnghiem_LSD_chuong_gemini.json'
OUTPUT_FILE = r'd:\Download\Thi\LSD\tracnghiem_LSD_chuong_gemini_fixed.json'
REPORT_FILE = r'd:\Download\Thi\LSD\bao_cao_loi.md'

print("Đang đọc file...")
with open(INPUT_FILE, 'r', encoding='utf-8') as f:
    data = json.load(f)

questions = data['questions']
total_original = len(questions)
print(f"Tổng số câu hỏi: {total_original}")

report = []
report.append("# BÁO CÁO PHÂN TÍCH & SỬA LỖI\n")
report.append(f"**File gốc:** `{os.path.basename(INPUT_FILE)}`\n")
report.append(f"**Tổng câu hỏi ban đầu:** {total_original}\n\n")

# ============================================================
# LỖI 5: OCR - chQ -> chủ, cQa -> của, phQ -> phủ, PhQ -> Phủ
# ============================================================
print("\n--- LỖI 5: Sửa OCR ---")
report.append("## 1. LỖI OCR (chQ → chủ, cQa → của, phQ → phủ)\n")

ocr_replacements = {
    'chQ': 'chủ',
    'cQa': 'của', 
    'phQ': 'phủ',
    'PhQ': 'Phủ',
}

ocr_fixes = []
for i, q in enumerate(questions):
    for field in ['question', 'answer']:
        original = q[field]
        fixed = original
        for old, new in ocr_replacements.items():
            fixed = fixed.replace(old, new)
        if fixed != original:
            ocr_fixes.append((i, field, original[:80], fixed[:80]))
            q[field] = fixed
    
    new_opts = []
    for j, opt in enumerate(q['options']):
        original = opt
        fixed = opt
        for old, new in ocr_replacements.items():
            fixed = fixed.replace(old, new)
        if fixed != original:
            ocr_fixes.append((i, f'option_{j}', original[:80], fixed[:80]))
        new_opts.append(fixed)
    q['options'] = new_opts

print(f"  Đã sửa {len(ocr_fixes)} lỗi OCR")
report.append(f"- **Đã sửa:** {len(ocr_fixes)} chỗ\n")
if ocr_fixes:
    report.append("- Chi tiết:\n")
    for idx, field, old_text, new_text in ocr_fixes[:20]:
        report.append(f"  - `[{idx}]` {field}: `{old_text}` → `{new_text}`\n")
    if len(ocr_fixes) > 20:
        report.append(f"  - ... và {len(ocr_fixes)-20} chỗ khác\n")
report.append("\n")

# ============================================================
# LỖI 1: Mất chữ đầu / Khuyết đoạn đầu
# ============================================================
print("\n--- LỖI 1: Mất chữ đầu ---")
report.append("## 2. MẤT CHỮ ĐẦU / KHUYẾT ĐOẠN ĐẦU\n")

lower_start_questions = []
digit_start_questions = []
for i, q in enumerate(questions):
    text = q['question'].strip()
    if not text:
        continue
    c = text[0]
    # Chữ cái thường ở đầu câu
    if c.islower() or c in 'àáảãạăắằẳẵặâấầẩẫậèéẻẽẹêếềểễệìíỉĩịòóỏõọôốồổỗộơớờởỡợùúủũụưứừửữựỳýỷỹỵđ':
        lower_start_questions.append((i, text[:120]))
    # Dính số ở đầu
    if c.isdigit():
        rest = text.lstrip('0123456789')
        if rest and rest[0] == ' ':
            digit_start_questions.append((i, text[:120]))

print(f"  Câu bắt đầu chữ thường: {len(lower_start_questions)}")
print(f"  Câu dính số ở đầu: {len(digit_start_questions)}")

report.append(f"### Câu bắt đầu bằng chữ thường: {len(lower_start_questions)}\n")
report.append("> ⚠️ Các câu này cần kiểm tra thủ công để bổ sung chữ/đoạn bị thiếu\n\n")
for idx, txt in lower_start_questions:
    report.append(f"- `[Câu {idx}]` {txt}\n")

report.append(f"\n### Câu dính số ở đầu: {len(digit_start_questions)}\n")
for idx, txt in digit_start_questions:
    report.append(f"- `[Câu {idx}]` {txt}\n")

# Tự động sửa: Xóa số dính ở đầu
for idx, _ in digit_start_questions:
    text = questions[idx]['question'].strip()
    questions[idx]['question'] = re.sub(r'^\d+\s+', '', text)

report.append(f"\n> ✅ Đã tự động xóa số dính ở đầu {len(digit_start_questions)} câu\n\n")

# ============================================================
# LỖI 2: Đáp án bị cắt xén
# ============================================================
print("\n--- LỖI 2: Đáp án bị cắt xén ---")
report.append("## 3. ĐÁP ÁN BỊ CẮT XÉN\n")

cut_words = [' và', ' của', ' tư', ' cho', ' để', ' là', ' về', ' theo', ' trong', 
             ' với', ' đã', ' một', ' các', ' có', ' được', ' do', ' từ', ' lên', 
             ' ra', ' vào', ' mà', ' này', ' đó', ' hay', ' hoặc']
cut_answers = []
for i, q in enumerate(questions):
    for j, opt in enumerate(q['options']):
        opt_clean = opt.strip()
        # Đáp án không kết thúc bằng ký tự hoàn chỉnh
        if opt_clean:
            last_char = opt_clean[-1]
            if last_char not in '.)?!…;:"\'':
                for cw in cut_words:
                    if opt_clean.endswith(cw):
                        cut_answers.append((i, j, chr(65+j), opt_clean, q['question'][:60]))
                        break

print(f"  Đáp án nghi bị cắt xén: {len(cut_answers)}")
report.append(f"- **Tổng nghi bị cắt:** {len(cut_answers)}\n")
report.append("> ⚠️ Cần kiểm tra thủ công, không thể tự đoán nội dung bị thiếu\n\n")
for idx, opt_idx, letter, opt_text, q_text in cut_answers:
    report.append(f"- `[Câu {idx}]` Đáp án **{letter}**: ...{opt_text[-60:]}\n")
    report.append(f"  - Câu hỏi: {q_text}\n")
report.append("\n")

# ============================================================
# LỖI 3: Thiếu phương án (< 4 đáp án)
# ============================================================
print("\n--- LỖI 3: Thiếu phương án ---")
report.append("## 4. THIẾU SỐ LƯỢNG PHƯƠNG ÁN (< 4)\n")

missing_opts = []
for i, q in enumerate(questions):
    n = len(q['options'])
    if n < 4:
        missing_opts.append((i, n, q['question'][:100]))

print(f"  Câu thiếu đáp án: {len(missing_opts)}")
report.append(f"- **Tổng:** {len(missing_opts)} câu\n")
report.append("> ⚠️ Cần bổ sung đáp án thủ công\n\n")
for idx, cnt, txt in missing_opts:
    report.append(f"- `[Câu {idx}]` Chỉ có **{cnt}** đáp án: {txt}\n")
report.append("\n")

# ============================================================
# LỖI 4: Sai format thứ tự phương án
# ============================================================
print("\n--- LỖI 4: Sai format thứ tự ---")
report.append("## 5. SAI FORMAT THỨ TỰ PHƯƠNG ÁN\n")

expected_letters = ['A', 'B', 'C', 'D']
wrong_order = []
for i, q in enumerate(questions):
    opts = q['options']
    letters = []
    for o in opts:
        o_s = o.strip()
        match = re.match(r'^([A-Z])\.\s', o_s)
        if match:
            letters.append(match.group(1))
        else:
            letters.append('?')
    expected = expected_letters[:len(opts)]
    if letters != expected:
        wrong_order.append((i, letters, expected, q['question'][:80]))

print(f"  Câu sai thứ tự: {len(wrong_order)}")
report.append(f"- **Tổng:** {len(wrong_order)} câu\n")
for idx, got, exp, txt in wrong_order:
    report.append(f"- `[Câu {idx}]` Có: {got}, Cần: {exp}\n")
    report.append(f"  - {txt}\n")

# Tự động sửa: đổi lại ký tự đầu cho đúng A, B, C, D
fixed_order = 0
for idx, got, exp, txt in wrong_order:
    opts = questions[idx]['options']
    new_opts = []
    old_answer = questions[idx]['answer']
    for j, opt in enumerate(opts):
        correct_letter = expected_letters[j]
        # Thay thế ký tự đầu
        new_opt = re.sub(r'^[A-Z]\.\s', f'{correct_letter}. ', opt.strip())
        # Nếu đáp án cũ chứa option text cũ, cập nhật
        if opt.strip() == old_answer.strip():
            questions[idx]['answer'] = new_opt
        new_opts.append(new_opt)
    questions[idx]['options'] = new_opts
    fixed_order += 1

report.append(f"\n> ✅ Đã tự động sửa thứ tự {fixed_order} câu\n\n")

# ============================================================
# LỖI 6: Câu hỏi lặp lại
# ============================================================
print("\n--- LỖI 6: Câu hỏi lặp ---")
report.append("## 6. CÂU HỎI BỊ LẶP LẠI\n")

q_texts = [q['question'].strip() for q in questions]
counter = Counter(q_texts)
duplicates = [(text, count) for text, count in counter.items() if count > 1]
duplicates.sort(key=lambda x: -x[1])
total_dup_extra = sum(c - 1 for _, c in duplicates)

print(f"  Nhóm câu lặp: {len(duplicates)}")
print(f"  Bản sao thừa: {total_dup_extra}")
report.append(f"- **Nhóm câu lặp:** {len(duplicates)}\n")
report.append(f"- **Bản sao thừa sẽ xóa:** {total_dup_extra}\n\n")
for text, count in duplicates[:30]:
    report.append(f"- [{count} lần] {text[:100]}\n")
if len(duplicates) > 30:
    report.append(f"- ... và {len(duplicates)-30} nhóm khác\n")

# Loại bỏ trùng lặp (giữ bản đầu tiên)
seen = set()
unique_questions = []
removed = 0
for q in questions:
    q_text = q['question'].strip()
    if q_text not in seen:
        seen.add(q_text)
        unique_questions.append(q)
    else:
        removed += 1

print(f"  Đã loại: {removed} câu trùng")
report.append(f"\n> ✅ Đã loại bỏ **{removed}** câu trùng lặp\n\n")

# ============================================================
# CẬP NHẬT VÀ GHI FILE
# ============================================================
data['questions'] = unique_questions
data['totalQ'] = len(unique_questions)

# Cập nhật section counts
section_counts = Counter(q['section'] for q in unique_questions)
for sec in data['sections']:
    sec['count'] = section_counts.get(sec['key'], 0)

report.append("---\n\n## TỔNG KẾT\n\n")
report.append(f"| Hạng mục | Số lượng |\n")
report.append(f"|----------|----------|\n")
report.append(f"| Câu hỏi ban đầu | **{total_original}** |\n")
report.append(f"| Câu sau khi loại trùng | **{len(unique_questions)}** |\n")
report.append(f"| Đã loại câu trùng | **{removed}** |\n")
report.append(f"| Sửa OCR | **{len(ocr_fixes)}** |\n")
report.append(f"| Sửa thứ tự đáp án | **{fixed_order}** |\n")
report.append(f"| Xóa số dính đầu | **{len(digit_start_questions)}** |\n\n")

report.append("### Phân bố theo chương:\n\n")
for sec in data['sections']:
    report.append(f"- **{sec['label']}**: {sec['count']} câu\n")

report.append("\n### Lỗi cần sửa thủ công:\n\n")
report.append(f"- Câu bắt đầu chữ thường (mất chữ đầu): **{len(lower_start_questions)}**\n")
report.append(f"- Đáp án nghi bị cắt xén: **{len(cut_answers)}**\n")
report.append(f"- Câu thiếu phương án: **{len(missing_opts)}**\n")

# Ghi file JSON đã sửa
with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)
print(f"\nĐã ghi file đã sửa: {OUTPUT_FILE}")

# Ghi báo cáo
with open(REPORT_FILE, 'w', encoding='utf-8') as f:
    f.writelines(report)
print(f"Đã ghi báo cáo: {REPORT_FILE}")

print(f"\n=== KẾT QUẢ ===")
print(f"Câu hỏi: {total_original} → {len(unique_questions)} (loại {removed} trùng)")
print(f"Sửa OCR: {len(ocr_fixes)}")
print(f"Sửa thứ tự: {fixed_order}")
print(f"Cần sửa thủ công: mất chữ đầu={len(lower_start_questions)}, cắt xén={len(cut_answers)}, thiếu đáp án={len(missing_opts)}")
